### PR TITLE
Final addition of shub endpoint functions

### LINKED ISSUE #434: add Singularity reponame/tag for shub:// endpoint
The [Singularity Hub API](https://singularity-hub.org/api/) endpoint is done to provide pulling `shub://` uri images based on the reponame and tag, eg:

https://singularity-hub.org/api/container/vsoch/singularity-images

or for a particular branch build:

https://singularity-hub.org/api/container/vsoch/singularity-images:master

and this of course needs to be added to the singularity command line application! I should get around to this in a few days, enough time for some folks to come back and merge other needed PRs :wink:

### PR DESCRIPTION
**Pull request by vsoch**
Fixes #434 #441 

This PR will add full functionality to all singularity hub endpoints (or ways to specify an image, including):

 - `shub://51` (unique id), the current implementation
 - `shub://vsoch/singularity-images` (namespace and repo name, tag defaults to master-->latest)
 - `shub://vsoch/singularity-images:latest` (specification of latest across all images)
 - `shub://vsoch/singularity-images:master` (specification of a tag)

I am going to wait to add official tests that pull images directly, as the database is going to be migrated to one more robust soon (and I will make permanent, official testing images at that time), but here is the functionality on command line:

Here is asking for an image based on username (akin to namespace) and repo name:

	vanessa@vanessa-ThinkPad-T460s:~/Desktop$ singularity shell shub://vsoch/singularity-images
	Cache folder set to /home/vanessa/.singularity/shub
	Found image vsoch/singularity-images:mongo
	Downloading image... 1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Decompressing /home/vanessa/.singularity/shub/1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Singularity: Invoking an interactive shell within container...
	Singularity.1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img> $ exit

Now we ask for the same image, but based on unique id:

	vanessa@vanessa-ThinkPad-T460s:~/Desktop$ singularity shell shub://55
	Cache folder set to /home/vanessa/.singularity/shub
	Found image vsoch/singularity-images:mongo
	Downloading image... 1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Decompressing /home/vanessa/.singularity/shub/1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Singularity: Invoking an interactive shell within container...
	Singularity.1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img> $ exit

Now we ask for a different image, the master branch (previous was branch "mongo")

	vanessa@vanessa-ThinkPad-T460s:~/Desktop$ singularity shell shub://vsoch/singularity-images:master
	Cache folder set to /home/vanessa/.singularity/shub
	Found image vsoch/singularity-images:master
	Downloading image... ce85ef57b5e9edf402f96b69d6ca44a6ce3adc91.img.gz
	Decompressing /home/vanessa/.singularity/shub/ce85ef57b5e9edf402f96b69d6ca44a6ce3adc91.img.gz
	Singularity: Invoking an interactive shell within container...
	Singularity.ce85ef57b5e9edf402f96b69d6ca44a6ce3adc91.img> $ exit

Asking for latest should return the latest image across tags (branches) which is mongo:

	vanessa@vanessa-ThinkPad-T460s:~/Desktop$ singularity shell shub://vsoch/singularity-images:latest
	Cache folder set to /home/vanessa/.singularity/shub
	Found image vsoch/singularity-images:mongo
	Downloading image... 1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Decompressing /home/vanessa/.singularity/shub/1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img.gz
	Singularity: Invoking an interactive shell within container...
	Singularity.1a4e2f28855f7e08dbe0fd969cb8372ffe005b4f.img> $ 

Changes proposed in this pull request

 - addition of all `shub://` uri endpoints so user's don't have to look up unique ids for images, and can specify easy-to-remember repo names. The previous specification of a particular image id is also still supported.
 - removing the redundant call to the docker API to retrieve the manifest twice, first to retrieve it, and then to get image with it.
 - Adding the parsing of the docker image name into a function, in anticipation of having command line completion (another PR).
- adding tests for the above functions

@singularityware-admin

### CODE REVIEW

**Comment by vsoch** [comment]
ok, ready for review!

File: libexec/python/utils.py [hunk 1]
```diff
@@ -172,19 +172,22 @@ def api_get(url,data=None,default_header=True,headers=None,stream=None,return_re
...
 
+
 def basic_auth_header(username, password):
     '''basic_auth_header will return a base64 encoded header object to
     generate a token
     :param username: the username
     :param password: the password
     '''
     s = "%s:%s" % (username, password)
+    logger.debug("AUTH: Input username/password: %s", s)
```
Comment (flx42) [inline libexec/python/utils.py]: Err, I don't think that's a good idea. If they have an issue, users will share theirs logs with us and they won't realize they are leaking their DockerHub password at the same time.

Comment (vsoch) [reply libexec/python/utils.py]: how would you suggest we add prints to debug this?

Comment (flx42) [reply libexec/python/utils.py]: We don't need to put debug statements everywhere!
This code is simple enough, if we fix the bug we currently have, I doubt we'll see any other bug related to this code in the future. So the debug code won't be needed. Well, hopefully.

Comment (vsoch) [reply libexec/python/utils.py]: oh, ok. I mentioned adding these in our discussion a few times, and you didn't say anything, so I thought that was the plan moving forward. I'm definitely glad to remove all prints and debug another way.

Comment (flx42) [reply libexec/python/utils.py]: Oh, I thought you meant to add them while investigating our authentication bug only. I didn't think you wanted to add the debug prints for everyone.

Comment (vsoch) [reply libexec/python/utils.py]: haha yeah np, I'm really unfortunately literal about most things :)

**Comment by vsoch** [comment]
@gmkurtzer this is ready for your review! The `shub://` tests have been added back with images that are intended to not go anywhere, and as we talked about, I changed the default behavior of "pull" to go to `PWD`. This was done by the addition of an argument for the client (`cli.py`) to take a `--pull-folder` and the pull.exec then sets this to be the `$PWD`. This could of course be changed to something else, if you carry forward some argument from singularity command line.

:D

File: libexec/cli/pull.exec [hunk 2]
```diff
@@ -86,7 +86,7 @@ case "$IMPORT_URI" in
     ;;
     shub://*)
         IMAGE_SHUB_ID=`echo "$IMPORT_URI" | sed -e 's@^shub://@@'`
-        exec $SINGULARITY_libexecdir/singularity/python/cli.py --shub $IMAGE_SHUB_ID
+        exec $SINGULARITY_libexecdir/singularity/python/cli.py --shub $IMAGE_SHUB_ID --pull-folder $PWD
```
Comment (gmkurtzer) [inline libexec/cli/pull.exec:89]: What does pull folder do? (just curious)?

Comment (gmkurtzer) [reply libexec/cli/pull.exec:89]: Ahh, looking at the API I think answered it for me. Where the image will land before it is run?

**Review by gmkurtzer — APPROVED** [review]
This looks good to me, no blockers that I can see so I will merge. Thanks!

File: libexec/cli/pull.exec (hunk 2 above)
Comment (vsoch) [reply libexec/cli/pull.exec:89]: `--pull-folder` is an argument that specifies the output directory for the pull. For the first PR it was just pulling (downloading) to the cache, but then you mentioned that it was more intuitive for it to plop wherever the user is, and I definitely agree. I hard coded the $PWD, but again this variable could be some higher level argument that the user can control (for example, if they want it to go to a cache, or to some folder for their images, etc.)

**Pull request merged**